src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs

```diff
@@ -406,19 +411,64 @@ private TypeSymbol BindTupleType(TupleTypeSyntax syntax, DiagnosticBag diagnosti
                 var argumentType = BindType(argumentSyntax.Type, diagnostics);
                 types.Add(argumentType);
 
-                // PROTOTYPE the error handling on missing, reserved or non-unique names should be fixed
-                var nameSyntax = argumentSyntax.Name;
-                var name = (object)nameSyntax != null ? nameSyntax.Identifier.ValueText : TupleTypeSymbol.UnderlyingMemberName(i);
-                names.Add(name);
+                string name = argumentSyntax.Name?.Identifier.ValueText;
+
+                // validate name if we have one
+                if (name != null)
+                {
+                    countOfExplicitNames++;
+
+                    int reserved = TupleTypeSymbol.IsMemberNameReserved(name);
+                    if (reserved == 0)
+                    {
+                        Error(diagnostics, ErrorCode.ERR_TupleReservedMemberNameAnyPosition, argumentSyntax.Name.Identifier, name);
+                    }
+                    else if (reserved > 0 && reserved != i + 1)
+                    {
+                        Error(diagnostics, ErrorCode.ERR_TupleReservedMemberName, argumentSyntax.Name.Identifier, name, i + 1);
+                    }
+                    else if (!uniqueFieldNames.Add(name))
+                    {
+                        Error(diagnostics, ErrorCode.ERR_TupleDuplicateMemberName, argumentSyntax.Name.Identifier);
+                    }
+                }
+
+                // add the name to the list
+                // names would typically all be there or none at all
+                // but in case we need to handle this in error cases
+                if (elementNames != null)
+                {
+                    elementNames.Add(name ?? TupleTypeSymbol.TupleMemberName(i));
+                }
+                else
+                {
+                    if (name != null)
```

Comment:
It looks like we have a test gap, this code is not reachable because we allocate elementNames builder at the point of  variable declaration. Please close the gap and adjust the code appropriately.